### PR TITLE
[Cache] Symfony PSR-6 implementation

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

From the provided README.md:

This component provides a strict [PSR-6](http://www.php-fig.org/psr/psr-6/) implementation for adding cache to your applications. It is designed to have a low overhead so that caching is fastest. It ships with a few caching adapters for the most widespread and suited to caching backends. It also provides a `doctrine/cache` proxy adapter to cover more advanced caching needs.

Todo:
- [ ] add more tests
- [ ] add a FilesystemAdapter

### CODE REVIEW

File: src/Symfony/Component/Cache/CacheItem.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function expiresAfter($time)
+    {
+        $now = time();
+
+        if (null === $expiration) {
```
Comment (vlaurier) [inline src/Symfony/Component/Cache/CacheItem.php]: I guess you wanted  to write $time, and not $expiration

**Comment by andrerom** [comment]
Interesting, like the simplistic approach, and especially that there is only one cache lookup for getItem call. But missing use of multi get / set / (...) to reduce roundtrips further. And calls to hasItem and getItem will result in two lookups here, but last point is maybe to be expected given the spec.

I know @Nyholm, @aequasi  and others have been [working on this topic]( https://github.com/php-cache), and hope to be able to contribute from eZ as well.

Would be good to join forces maybe.

<hr>

Side: Personally I would go for:
- two layers, the service (api) psr6 layer, and the adapter/driver layer with more low level interface(s)
 - Avoids having to use extends and ever growing abstract adapter base class
 - Allows people that only want the key/value to use the low level at own risk.
 - Be able to add features and change interfaces in low level without affecting psr6 api, also be able to adopt future changes from PSR w/o potential conflicts.
- Fork or re factor *in sync with the given project* an existing adapter library that supports:
 -  Multi get / set / .. => to be able to do as few calls to cache backend as possible, to reduce server roundtrips and thus application latency
 - Already supporting lots of cache backends with full test suites  => avoid redoing bugs
 - This could be `doctrine/cache`, with cleanup of structure and getting rid of extra cache lookup for namespace in favor of approach here, however this would be needed to be discussed with @Ocramius and others.
 - Personally I also strongly think it should support Tagging* quite early on, or easily be able to add it, this can optionally be used by end users to solve the use case behind the doctrine cache namespace => clear all


* TL;DR; Tagging is a superior cache expiry strategy for any full scale applications, where a entity almost always is used several places that is cached, and tagging allows you to keep track of this to expiry reliably on a given entity change. *See for instance @magnusnordlander's talk from Symfony Con., talks / blogs on Drupal8 cache impl., or talks on FOSHttpCache*

**Comment by Nyholm** [comment]
Thank you for the ping.

I ran this implementation on our [integration tests](https://github.com/php-cache/integration-tests) and found some issues. Like you do not protect for all the reserved characters and you have forgotten to clear the deferred items on `$pool->clear()`.

The [php-cache](http://www.php-cache.com) supports tagging, hierarchal keys and have doctrine bridges in both directions. We are currently outperforming other PSR-6 caches with a factor of 2 and we have ideas for even more performance improvements. 

We would definitely like to join forces somehow. We are open for a discussion about that.

File: src/Symfony/Component/Cache/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Symfony PSR-6 implementaton for caching
```
Comment (pborreli) [inline src/Symfony/Component/Cache/README.md]: implementaton  => implementation

File: src/Symfony/Component/Cache/CacheItem.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        $now = time();
+
+        if (null === $time) {
+            $this->lifetime = $this->defaultLifetime;
+        } elseif ($time instanceof \DateInterval) {
+            $time = \DateTime::createFromFormat('U', $now)->add($time)->format('U');
+        } elseif (is_int($time)) {
+            $time += $now;
+        } else {
+            throw new InvalidArgumentException(sprintf('Expiration date must implement DateTimeInterface or be null, "%s" given', is_object($time) ? get_class($time) : gettype($time)));
```
Comment (sstok) [inline src/Symfony/Component/Cache/CacheItem.php]: Exception message is wrong

**Comment by nicolas-grekas** [comment]
Comments addressed, thank you @Nyholm for the link to the integration test, this will definitely help.
As I tried to state in the README, I target a strict implementation of PSR-6 here. The goal is not to replace any existing caching libraries (php-cache, stash, doctrine/cache, etc.) but only provide the features that haven be standardized, so that we can build interoperability on top. This means that adding more interfaces of features (tagging) is out of scope, unless a new PSR standardize them.

**Comment by nicolas-grekas** [comment]
I added testing with cache/integration-tests, I'll look at the failures asap. I already fixed the missing `:` reserved char.

**Comment by Nyholm** [comment]
> I target a strict implementation of PSR-6 here. The goal is not to replace any existing caching libraries (php-cache, stash, doctrine/cache, etc.) but only provide the features that have be standardized, so that we can build interoperability on top. 

If the goal is not to replace any existing cache libraries, why build another cache library? Wouldn't it be better if all Symfony components using cache relied on the PSR6 interface and then required the virtual package `psr/cache-implementation: ~1.0`?

That way we use the benefit of PSRs and we can let the user decide what cache library to use. If they like to use feature X they install a library with feature X. If they want a strict implementation they would choose a strict implementation.

**Comment by fabpot** [comment]
@Nyholm ... and this component is a strict implementation that users can use if they want to, like any other ones. The goal of PSRs (if I get this right) is to allow several implementations with different philosophies, not to standardize on only one.

**Comment by Nyholm** [comment]
> @Nyholm ... and this component is a strict implementation that users can use if they want to, like any other ones. The goal of PSRs (if I get this right) is to allow several implementations with different philosophies, not to standardize on only one.

You are correct. That is the goal of PSRs. But it was this sentence I reacted to: 

> The goal is not to replace any existing caching libraries

If the goal **is** to provide a strict implementation of PSR-6 that **will** replace (or be an option to) any existing library, then I agree that we need this component. Otherwise I do not see the benefits at all.

**Comment by fabpot** [comment]
Understood, we definitely want to provide a strict implementation that **can** replace existing libraries :)

**Comment by Nyholm** [comment]
Okey, awesome. 

Im happy to help. @nicolas-grekas do you want PR's with more adapters now or once this is merged?

**Comment by nicolas-grekas** [comment]
@Nyholm thank you that's much appreciated. Your test suite is already an amazing help. I think it' would be easier to add more adapters after this one has been merge.
For now I have the following failures (see below).

> 1) Symfony\Component\Cache\Tests\CacheTest::testSaveWithNoData
> When saving an object with no data, we should generate an error.

I didn't read anything about this in PSR-6, did I miss something?

For the other ones, I'll check tomorrow hopefully:

> 2) Symfony\Component\Cache\Tests\CacheTest::testDeferredSaveWithoutCommit
> A deferred item should automatically be committed on CachePool::__destruct().

> 3) Symfony\Component\Cache\Tests\CacheTest::testExpiration
> Failed asserting that true is false.

> 4) Symfony\Component\Cache\Tests\CacheTest::testIsHit
> Failed asserting that false is true.

> 5) Symfony\Component\Cache\Tests\CacheTest::testIsHitDeferred
> Failed asserting that false is true.

**Comment by dunglas** [comment]
Using a standardized interface for cache in Symfony will be very helpful!

What do you think about using the PSR-6 interface instead of the Doctrine Cache one in #16838 and #16917?

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        return $ok;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(CacheItemInterface $item)
+    {
+        if (!$item instanceof CacheItem) {
```
Comment (blanchonvincent) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: Why do you need to do this?
It breaks the usage of `CacheItemInterface` interface used as type hinting.

**Comment by Nyholm** [comment]
1) When saving an object with no data, we should generate an error.

We have given some thoughts in to this. The PSR doc block on `save` says "*True if the item was successfully persisted. False if there was an error.*". I claim it is an error because `null` is valid data. 

Consider this: 
```php
$item = $pool->getItem('key');
$pool->save($item); // Assume this is okey

$pool->hasItem('key'); // Should return true, yes? Or else, what is the point of saving?

/** 
 * Doc block for get(): 
 * If isHit() returns false, this method MUST return null. Note that null
 * is a legitimate cached value, so the isHit() method SHOULD be used to
 * differentiate between "null value was found" and "no value was found."
 * 
 * Doc block for isHit():
 * Confirms if the cache item lookup resulted in a cache hit.
 */
$pool->getItem('key')->isHit(); // Must be false because "no value was found."

// So ...
$pool->hasItem('key') !== $pool->getItem('key')->isHit(); // This makes no sense.
```

The pool should not have an item that never can be a hit. That would be a stale item and should be removed.
If you really want to store just the key you should just do `$item->set(null);`

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 4 above)
Comment (xavierleune) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: @blanchonvincent the PSR specifies that implementation are not compatible. So the type hinting is specified by the PSR interface, but the $item MUST be an instance from the same implementation (i.e: I cannot make work my CacheItem from my custom Redis implementation to work with your custom Memcached implementation of CacheItemPool).

**Comment by nicolas-grekas** [comment]
@Nyholm I don't get your example, to me, assuming "key" is not already in the cache:
```php
$pool->hasItem('key'); // false
$item = $pool->getItem('key');
$item->get(); // null
$item->isHit(); // false
$pool->save($item); // true
$item = $pool->getItem('key');
$item->get(); // null
$item->isHit(); // true -> a value has been found, we saved it before!
```

**Comment by Nyholm** [comment]
You have not saved any value. Since: "*Note that null is a legitimate cached value, so the isHit() method SHOULD be used to differentiate between "null value was found" and "no value was found."*"

The last two rows in you example: 
```php
$item->get(); // null
$item->isHit(); // true -> a value has been found, we saved it before!
```
If `$item->isHit() === true` and `$item->get()` returns `null` you assume that someone has done a `$item->set(null)` which is not the case. 

The point is, `null` should not be treated any special.

**Comment by nicolas-grekas** [comment]
> If $item->isHit() === true and $item->get() returns null you assume that someone has done a $item->set(null) which is not the case. 

My example above clearly allows to differentiate between a null a "found" vs a null as "not found". Nothing it the spec says that `$item->set(null)` is the only way to get a "data" null. This still looks like a non standard requirement to me.

**Comment by Nyholm** [comment]
If you do not agree with it, [mark it as skipped](https://github.com/php-cache/void-adapter/blob/master/tests/IntegrationPoolTest.php) for now and we can have the discussion about it on the integration test repo. 

I think it make sense but I will ask the in the mailing list what the intention might have been and adjust the integration tests accordingly. 

Thank you for your input.

**Comment by nicolas-grekas** [comment]
@Nyholm discussion about conformance tests can continue here: https://github.com/php-cache/integration-tests/pull/33

**Comment by fabpot** [comment]
:+1: ping @symfony/deciders

**Comment by dunglas** [comment]
:+1:

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = array())
+    {
+        $f = $this->createCacheItem;
+        $items = array();
+        $now = time();
+
+        foreach ($keys as $key) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: Just asking: could we use PHP generators in loops like this one?

```diff
foreach ($keys as $key) {
    $isHit = isset($this->expiries[$this->validateKey($key)]) && ($this->expiries[$key] >= $now || !$this->deleteItem($key));           
-    $items[$key] = $f($key, $isHit ? $this->values[$key] : null, $isHit);
+    yield $key => $f($key, $isHit ? $this->values[$key] : null, $isHit);
}

-return $items;
```

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 4 above)
Comment (gcds) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: You breaking the whole idea of PSR-6 you can interchange multiple libraries together... That's why Interfaces exists

**Comment by Nyholm** [comment]
I do also think this looks good. :+1:

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: @gcds did you read @xavierleune just in top of yours?

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 5 above)
Comment (Nyholm) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: Sorry no, `getItems` MUST return an `array` or `\Traversable`.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 4 above)
Comment (Nyholm) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: From the [PSR](http://www.php-fig.org/psr/psr-6/) itself: 

> Calling Libraries SHOULD NOT assume that an Item created by one Implementing Library is compatible with a Pool from another Implementing Library.

Comment (gcds) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: Ahh, sorry I thought differently...

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 5 above)
Comment (gcds) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: @Nyholm if I remember correctly Generator implements Iterator (Iterator extends Traversable)

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: I don't think a generator would provide an benefit here. It could even add a bit of overhead.
If you really think it's better, I suggest you open a PR after this one has been merged.

**Comment by nicolas-grekas** [comment]
@Nyholm the only thing that is missing to turn tests green is a merge of https://github.com/php-cache/integration-tests/pull/33 :)
Status: reviewed

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: BTW, Interoperability between implementations is possible using something like the attached ProxyAdapter

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 5 above)
Comment (gcds) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: If you think generator are bad go read about them... Generators save memory and other good stuff

File: composer.json [hunk 6]
```diff
@@ -74,6 +76,8 @@
         "symfony/yaml": "self.version"
     },
     "require-dev": {
+        "cache/integration-tests": "dev-master",
```
Comment (Nyholm) [inline composer.json]: Use version `^0.6` 

A PR to this repo should not fail because we added extra tests. IMHO it is better to make a PR to update the version of the integration tests now and then.

**Comment by nicolas-grekas** [comment]
Tests are greenn thank you @Nyholm 
I suggest merging this one asis and doing any further enhancements on next PRs (esp. adding adapters)

**Comment by Nyholm** [comment]
Thank you for your feedback on the tests. 
I agree. This should be merged now.

**Comment by JHGitty** [comment]
@nicolas-grekas Thank you for working on this - it is awesome!

**Comment by dunglas** [comment]
+1 to merge it now. It will allow to start migrating metadata systems and other places in Symfony using Doctrine Cache internally to PSR-6.

File: src/Symfony/Component/Cache/composer.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        }
+    ],
+    "require": {
+        "php": ">=5.5.9",
+        "psr/cache": "~1.0"
+    },
+    "require-dev": {
+        "cache/integration-tests": "^0.6",
+        "doctrine/cache": "~1.6"
+    },
```
Comment (joelwurtz) [inline src/Symfony/Component/Cache/composer.json]: I think it miss

```
provides: {
    "psr/cache-implementation": "1.0"
},
```

?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/composer.json]: @joelwurtz indeed, fixed thanks

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by Tobion** [comment]
IMO namespaces handling should be a separate decorator and not built into the adapter directly. See https://github.com/doctrine/cache/pull/96

**Comment by Tobion** [comment]
I also want to express my objections against the recent fast merges of new components. This was merged in less than two days which does not give enough time to review and discuss a full new component at all. Also it's against our own rules which says

> A pull request can be merged if: Enough time was given for peer reviews (at least 2 days for "regular" pull requests, and 4 days for pull requests with "a significant impact");
http://symfony.com/doc/current/contributing/code/core_team.html

I hope it's not going to be of the same quality as the LDAP component.

**Comment by andrerom** [comment]
> IMO namespaces handling should be a separate decorator and not built into the adapter directly. See doctrine/cache#96

True, but also a bit of a orange to apple comparison. In doctrine this is a namespace kept in cache and that needs to be checked on every cache lookup, that is not the case here.

**Comment by nicolas-grekas** [comment]
@Tobion sorry if you feel uncomfortable, we merged quickly that's true... I invite you to open a PR to move the namespace concept in it's own decorator if you think it's better (taking @andrerom into account). The component will see more enhancements until 3.1 is released, but at least we don't have to discuss about it's interfaces, which makes the situation totally different from the Ldap component hopefully.

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        return $this->pool->commit();
+    }
+
+    private function doSave(CacheItemInterface $item, $method)
+    {
+        if (!$item instanceof CacheItem) {
+            return false;
+        }
+        static $prefix = "\0Symfony\Component\Cache\CacheItem\0";
```
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: Why not a `const`?

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ProxyAdapter implements CacheItemPoolInterface
```
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: It deserves a little description of the usage IMO.

`The ProxyAdapter class allows to use another cache item pool that complies with PSR-6 with the Symfony Cache component.`

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    private function doSave(CacheItemInterface $item, $method)
+    {
+        if (!$item instanceof CacheItem) {
+            return false;
+        }
+        static $prefix = "\0Symfony\Component\Cache\CacheItem\0";
+        $item = (array) $item;
+        $poolItem = $this->pool->getItem($item[$prefix.'key']);
```
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: It will throw an error because `\0` isn't an allowed character in keys according to PSR-6.

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ArrayAdapter implements CacheItemPoolInterface
```
Comment (andrerom) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: @nicolas-grekas Before considering to open pr on documenting this adapter better, what is your's intention behind this one?

A. Do you see it only for testing, hence we should clarify and avoid using it in doc examples. 
B. Or do you think it should be supported in prod using a chain proxy in front of an actual cache pool for performance? Meaning it will need to handle limits to avoid using all memory, and will also need to reduce risk of returning stale data while also documenting that there will always be a risk of returning stale data *(unless we add flag to getItems to on a case by case basis let users specify if they accept stale data or not, default being not)*.

Context:
Stash has issues with this, the adapter is php documented as for unit tests only, but the bundle exposes an option to enable "InMemory" cache, w/o further details on the risk of broken application logic it might cause.
See alos: https://github.com/php-cache/array-adapter/issues/3

**Comment by gcds** [comment]
My suggestion for namespace would be to let Adaptor generate key with a namespace. For e.g. most Redis tools groups keys if they have `:` separators. If Adaptor for e.g. doesn't support some characters or uses own rules with namespaces it would be better I think in this way. Like I said for Redis e.g. to replace `\` with `:` it would allow better viewing for redis tools...

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php (hunk 8 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: because a const is public...

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php (hunk 9 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: I added a small one in the README, could be better (and needs doc on symfony-doc :) )

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php (hunk 10 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: you miss read: the $prefix is used locally to access the array cast, it's not given to the pool

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php (hunk 8 above)
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: `@internal`? Or a private static property? This one looks a bit weird.

File: src/Symfony/Component/Cache/Adapter/ProxyAdapter.php (hunk 10 above)
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/ProxyAdapter.php]: Ok got it.

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: @andrerom Your concern is filling memory with stale data or something more? There is no risk to return stale data, since the expiry is verified. The use case I see is when a non persistent in memory cache is enough. Could be testing or something else. We could enhance the logic for dealing with expirations, such has maintaining a property that stores the next full scan for expired items, to be checked e.g. on every access. Would you like to work on it?

**Comment by nicolas-grekas** [comment]
@gcds namespace (or prefix really) are already managed by adapters (they must deal with it in their constructor). Which means a Redis adapter could do whatever you'd need there. To me, the simple prefixing logic is rightfully in the abstract, because adding a prefix is really something common. Doing namespaces àla doctrine/cache is for a decorator, or at least not for the abstract, sure.

**Comment by gcds** [comment]
Ahh yeah, sounds correct to me :)

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 11 above)
Comment (andrerom) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: Two issues:
- There is no limit here so you run out of memory on long running scripts, could be a configurable argument with default set to 100 for instance
- If you aim to support this for setups where there is any concurrence *(several requests/processes/servers)*, then the more concurrency you have + the longer your script runs, the higher likelihood of ending up getting stale data back from cache as local cache doesn't get invalidated.

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: I don't get why you mean by stale data: to me, data-staleness in the context of PSR-6 is defined by the expiry date. Until this date is reached, the data is not stale. After, it is. ArrayAdapter is not different from any other adapter when using this definition. The only concern I see that is specific to ArrayAdapter is garbage collecting the stale items. Do I miss something in what you said?

Comment (andrerom) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: Expiry alone is worthless if you for instance use the cache service in a decorator for your persistence / data store ( [Example](https://doc.ez.no/display/TECHDOC/Persistence+cache) ), in such a scenario you need to invalidate cache on changes to the data.

Feel free to suggest a place to discuss this.

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: a better place would be a new issue here on github :)

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+    private $namespace;
+    private $deferred = array();
+    private $createCacheItem;
+    private $mergeByLifetime;
+
+    protected function __construct($namespace = '', $defaultLifetime = 0)
+    {
+        $this->namespace = $namespace;
+        $this->createCacheItem = \Closure::bind(
+            function ($key, $value, $isHit) use ($defaultLifetime) {
```
Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: What about using adding a constructor to `CacheItem` marked with `@internal`? It will be easier to understand. (Currently even PHPStorm isn't able to interpret this snippet and marks it in red).

Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: Btw it will remove the duplication of this factory closure in `ProxyAdapter`.

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: I preferred raw speed over phpstorm compliance: a constructor would slow down things for no valid reason (adding one function call in the chain). I doubt `@internal` would be enough to prevent erroneous user side instantiation. The current no-constructor-nor-setter (for key and isHit) means we technically annoy users enough so that they are free to missuse but forced to thing about it :)

Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: Isn't a constructor call or a closure call in the chain similar in term of perfs?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: yes of course, but making the constructor public has drawback IMHO (as detailed above. This drawback could be fixed by using a private constructor + a rebound closure, which when an additional function call comes up. That's what I meant :)

File: src/Symfony/Component/Cache/Adapter/ArrayAdapter.php (hunk 11 above)
Comment (andrerom) [inline src/Symfony/Component/Cache/Adapter/ArrayAdapter.php]: Here you go: https://github.com/symfony/symfony/issues/17445
